scheduler/servicestatemanager/servicestatemanager.go

```diff
@@ -0,0 +1,947 @@
+	// AddTenant prepares the service state manager to receive requests for a new tenant
+	AddTenant(tenantID string) error
+	// RemoveTenant notifies the service state manager that a tenant no longer exists
+	RemoveTenant(tenantID string) error
+	// WaitScheduled blocks until all requested services have been scheduled or cancelled
+	WaitScheduled(tenantID string, serviceIDs ...string)
+	// Wait blocks until all processing for the current tenant has completed
+	Wait(tenantID string)
+}
+
+// ServiceStateChangeBatch represents a batch of services with the same
+// desired state that will be operated on by a ServiceStateManager
+type ServiceStateChangeBatch struct {
+	Services     map[string]CancellableService
+	DesiredState service.DesiredState
+	Emergency    bool
+}
+
+// CancellableService is a service whose scheduling may be canceled by a channel
+type CancellableService struct {
+	*service.Service
+	cancel chan interface{}
+	C      <-chan interface{}
+}
+
+// Types for sorting
+type CancellableServices []CancellableService
+
+func (s CancellableServices) Len() int      { return len(s) }
+func (s CancellableServices) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+
+type ByEmergencyShutdown struct{ CancellableServices }
+
+// Sort by EmergencyShutdownLevel - 1, to ensure level of 0 is last
+func (s ByEmergencyShutdown) Less(i, j int) bool {
+	if s.CancellableServices[i].EmergencyShutdownLevel == s.CancellableServices[j].EmergencyShutdownLevel {
+		// If emergency shutdown level is the same, order by reverse start level
+		return s.CancellableServices[i].StartLevel-1 > s.CancellableServices[j].StartLevel-1
+	} else {
+		return s.CancellableServices[i].EmergencyShutdownLevel-1 < s.CancellableServices[j].EmergencyShutdownLevel-1
+	}
+}
+
+type ByStartLevel struct{ CancellableServices }
+
+func (s ByStartLevel) Less(i, j int) bool {
+	return s.CancellableServices[i].StartLevel-1 < s.CancellableServices[j].StartLevel-1
+}
+
+type ByReverseStartLevel struct{ CancellableServices }
+
+func (s ByReverseStartLevel) Less(i, j int) bool {
+	return s.CancellableServices[i].StartLevel-1 > s.CancellableServices[j].StartLevel-1
+}
+
+// ServiceStateQueue is a queue with a queueLoop processing it's batches
+type ServiceStateQueue struct {
+	sync.RWMutex
+	BatchQueue   []ServiceStateChangeBatch
+	CurrentBatch ServiceStateChangeBatch
+	Changed      chan bool
+	Facade       Facade
+}
+
+// BatchServiceStateManager intelligently schedules batches of services to start/stop/restart with the facade
+type BatchServiceStateManager struct {
+	sync.RWMutex
+	Facade                 Facade
+	ctx                    datastore.Context
+	ServiceRunLevelTimeout time.Duration
+	TenantQueues           map[string]map[service.DesiredState]*ServiceStateQueue
+	TenantShutDowns        map[string]chan<- int
+}
+
+func NewCancellableService(svc *service.Service) CancellableService {
+	cancel := make(chan interface{})
```

Comment:
Unfortunately, that eventually gets passed to zzk.WaitService, which requires a chan interface{}.